### PR TITLE
Replace device_vector with device_uvector in sort_impl

### PR DESCRIPTION
**Pull request by davidwendt**
Reference #7287

This replaces usages of `rmm::device_vector` with `rmm::device_uvector` in `cpp/src/sort/sort_impl.cuh` which is used internally by the `cudf::sort` APIs.
The `make_device_uvector_async` utility are called to convert a `std::vector` to a temporary `rmm::device_uvector`.

Also updated `vector_factories.hpp` utility to add the missing `include <vector>`.

### CODE REVIEW

File: cpp/src/sort/sort_impl.cuh [hunk 1]
```diff
@@ -123,14 +125,14 @@ std::unique_ptr<column> sorted_order(table_view input,
...
+  auto device_table         = table_device_view::create(input_flattened, stream);
+  auto const d_column_order = make_device_uvector_async(std::get<1>(flattened), stream);
 
   if (has_nulls(input_flattened)) {
-    rmm::device_vector<null_order> d_null_precedence(std::get<2>(flattened));
-    auto comparator = row_lexicographic_comparator<true>(
-      *device_table, *device_table, d_column_order.data().get(), d_null_precedence.data().get());
+    auto d_null_precedence = make_device_uvector_async(std::get<2>(flattened), stream);
+    auto comparator        = row_lexicographic_comparator<true>(
+      *device_table, *device_table, d_column_order.data(), d_null_precedence.data());
```
Comment (ttnghia) [inline cpp/src/sort/sort_impl.cuh:135]: Should `d_null_precedence ` and `comparator` be `const`?

File: cpp/src/sort/sort_impl.cuh [hunk 2]
```diff
@@ -142,9 +144,11 @@ std::unique_ptr<column> sorted_order(table_view input,
                    mutable_indices_view.end<size_type>(),
                    comparator);
     }
+    // protection for temporary d_column_order and d_null_precedence
+    stream.synchronize();
```
Comment (ttnghia) [inline cpp/src/sort/sort_impl.cuh:148]:
```suggestion
```

File: cpp/src/sort/sort_impl.cuh [hunk 3]
```diff
@@ -156,6 +160,8 @@ std::unique_ptr<column> sorted_order(table_view input,
                    mutable_indices_view.end<size_type>(),
                    comparator);
     }
+    // protection for temporary d_column_order
+    stream.synchronize();
   }
```
Comment (ttnghia) [inline cpp/src/sort/sort_impl.cuh:165]:
```suggestion
}

// protection for temporary d_column_order
stream.synchronize();
```

The `stream.synchronize();` statement is called at the end of each blocks `if else`, thus we can move it out of those blocks.

File: cpp/src/sort/sort_impl.cuh (hunk 2 above)
Comment (davidwendt) [reply cpp/src/sort/sort_impl.cuh:148]: The `d_null_precedence` temporary goes out of scope and gets destroyed here.
So this sync is required inside this scope.
The comment was meant to convey that information.

**Review by nvdbaranec — APPROVED** [review]

**Review by ttnghia — APPROVED** [review]

File: cpp/include/cudf/detail/utilities/vector_factories.hpp [hunk 4]
```diff
@@ -28,6 +28,8 @@
 #include <rmm/device_uvector.hpp>
 #include <rmm/mr/device/device_memory_resource.hpp>
 
+#include <vector>
```
Comment (cwharris) [inline cpp/include/cudf/detail/utilities/vector_factories.hpp:31]: is this needed?

**Review by cwharris — APPROVED** [review]

File: cpp/include/cudf/detail/utilities/vector_factories.hpp (hunk 4 above)
Comment (davidwendt) [reply cpp/include/cudf/detail/utilities/vector_factories.hpp:31]: There are functions in this header that create and return `std::vector`.
The `vector_factories.hpp` should not rely on other files to include dependent headers for it.
I hit a compile error for this on some prototype test code that included `vector_factories.hpp` but did not need the `std::vector` functions.

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**